opengrok-indexer/src/test/java/org/opengrok/indexer/util/ProgressTest.java

```diff
@@ -0,0 +1,90 @@
+import static org.mockito.Mockito.atLeast;
+import static org.mockito.Mockito.times;
+
+public class ProgressTest {
+    @BeforeClass
+    public static void setup() {
+        // needed to spawn logger thread in Progress
+        RuntimeEnvironment.getInstance().setPrintProgress(true);
+    }
+
+    @Test
+    public void testProgress() throws InterruptedException {
+        final Logger logger = Mockito.mock(Logger.class);
+        final int totalCount = 100; // use int because of Mockito.calls()/times() only accepts int
+
+        Mockito.when(logger.isLoggable(any())).thenReturn(true);
+        try (final Progress progress = new Progress(logger, "foo", totalCount)) {
+            assertNotNull(progress.getLoggerThread());
+
+            // Progress does not give any indication that the logger thread reached the point when
+            // it can accept notifications from increment() so check the Thread state.
+            while (progress.getLoggerThread().getState() != Thread.State.WAITING) {
+                System.out.println("Waiting for the logger thread to reach the initial wait()");
+                TimeUnit.MILLISECONDS.sleep(10);
+            }
+
+            for (int i = 0; i < totalCount; i++) {
+                progress.increment();
+                // Give the logger thread some time to log.
+                while (progress.getLoggerThread().getState() != Thread.State.WAITING) {
+                    TimeUnit.MILLISECONDS.sleep(10);
+                }
+            }
+        }
+
+        Mockito.verify(logger, times(totalCount)).log(any(), anyString(), any(Object[].class));
+    }
+
+    @Test
+    public void testThreads() throws InterruptedException {
+        final Logger logger = Mockito.mock(Logger.class);
+        final int totalCount = Runtime.getRuntime().availableProcessors();
+        List<Future<?>> futures = new ArrayList<>();
+
+        Mockito.when(logger.isLoggable(any())).thenReturn(true);
+        ExecutorService executor = Executors.newFixedThreadPool(totalCount);
+        System.out.println(String.format("Will run %d threads", totalCount));
+        try (final Progress progress = new Progress(logger, "foo", totalCount)) {
+            while (progress.getLoggerThread().getState() != Thread.State.WAITING) {
+                System.out.println("Waiting for the logger thread to reach the initial wait()");
+                TimeUnit.MILLISECONDS.sleep(10);
+            }
+
+            for (int i = 0; i < totalCount; i++) {
+                futures.add(executor.submit(progress::increment));
+            }
+        }
+
+        System.out.println("Waiting for threads to finish");
+        futures.forEach(future -> {
+            try {
+                future.get();
+            } catch (Exception e) {
+                Assert.fail();
+            }
+        });
+        executor.shutdown();
+        System.out.println("Verifying");
+        Mockito.verify(logger, atLeast(1)).log(any(), anyString(), any(Object[].class));
```

Comment:
If the futures waits are changed to happen before `close()`, might this be strictly `times(totalCount)` as with the previous test?

Reply:
this case does not wait for the logger thread to reach the waiting state so it usually is less that total count.